cts/cli/regression.daemons.exp

```diff
@@ -101,52 +101,52 @@
         <option value="panic"/>
       </content>
     </parameter>
-    <parameter name="election-timeout">
+    <parameter name="election-timeout" advanced="true" generated="false">
       <longdesc lang="en">
         Declare an election failed if it is not decided within this much time. If you need to adjust this value, it probably indicates the presence of a bug.
       </longdesc>
       <shortdesc lang="en">
-        *** Advanced Use Only ***
+        Declare an election failed if it is not decided within this much time. If you need to adjust this value, it probably indicates the presence of a bug.
```

Comment:
> check for

All `pcs` or `crmsh` would have to do is look for `--list-options` in the output of `crm_attribute --help-all`. If and only if it's present, you can run `crm_attribute --list-options=cluster --output-as=xml` to get all the options in the new format instead of running the daemon metadata commands.

Whether you all have time to add parsing for the new format, is another matter.

> I'm guessing with current pcs code, we would just lose the indicators

It actually breaks entirely.
```
$ sudo pcs property describe
Error: Agent 'pacemaker-based' is not installed or does not provide valid metadata: Invalid attribute advanced for element parameter, line 12
Error: Agent 'pacemaker-controld' is not installed or does not provide valid metadata: Invalid attribute advanced for element parameter, line 12
Error: Agent 'pacemaker-schedulerd' is not installed or does not provide valid metadata: Invalid attribute advanced for element parameter, line 12
Error: Errors have occurred, therefore pcs is unable to continue
```

pcs is trying to validate the output against the OCF RA 1.1 schema. The OCF schema doesn't allow these custom attributes, which is why I've overridden a few pieces of it in the new `crm_attribute-2.34.rng` schema.

---

Incidentally, pcs's parsing and of allowed values is broken since 2.1.7. It very specifically depends on the presence of TWO spaces before "Allowed values" in the longdesc. But in 2.1.7, I removed one of the spaces as gratuitous, assuming it didn't matter.
```
lib/resource_agent/pcs_transform.py:        parts = parameter.longdesc.split("  Allowed values: ")
```

Reply:
@kgaillot @nrwahl2:
Validating against modified OCF schema (`/usr/lib/pcs/data/ocf-1.1.rng`) is hard-coded and it cannot go away. The reason is that the real OCF schema has never been actually adhered to. To be able to at least somehow validate the xml structure, we had to come up with a relaxed version of the schema and bundle it with pcs.

Moving from plaintext to structured data for marking advanced parameters, allowed parameter values and deprecations is welcome. However, pcs needs to be compatible with both the new and the old way at the same time. We cannot be sure that upstream users match the new pcs with the new pacemaker. It would be easier to have OCF 1.2 standard to differentiate between the two approaches (plaintext vs. structure). (And to prevent further deviation from the OCF 1.1.) But I think we can manage it in pcs with OCF 1.1 as well, as long as there is either of plaintext or attribute specification and not both for the given metadata. If you want this to be done in 9.5, it is manageable. It would certainly help to finish the pacemaker side sooner than later, so we have enough time to get it done.

> Incidentally, pcs's parsing of allowed values is broken since 2.1.7. It very specifically depends on the presence of TWO spaces before "Allowed values" in the longdesc.

That's easy to fix. Just switch such parameters to `content type="select"` and put the allowed values to `option` elements, as specified in OCF 1.1 :wink: 

Moving to `crm_attribute --list-options=cluster` is another change on top of that, though. If the old way of running daemon metadata commands is preserved, no matter if it produces metadata in the current format or the one discussed here, I'm fine with that. It would give us time to switch to `crm_attribute`.

Reply:
@tomjelinek
> Validating against modified OCF schema (`/usr/lib/pcs/data/ocf-1.1.rng`) is hard-coded and it cannot go away.

All we'd have to do is update that schema to allow two new optional attributes in the `<parameter>` element: `advanced` and `generated`. Since they're optional, it wouldn't break on older Pacemakers.

> OCF 1.2 standard to differentiate between the two approaches

I have to defer to Ken on whether it's reasonable to have a new official OCF schema accept this output. The three differences between this cluster option output and the OCF 1.1 schema are:
* No `<actions>` element (required by OCF)
* Two option types (`percentage` and `time`) that are not allowed by OCF
* Two new `<parameter>` attributes (`advanced` and `generated`)

Depends on whether we want to relax the official schema in those ways. It *would* be backward compatible, since nothing that was previously accepted would become rejected.

> as long as there is either of plaintext or attribute specification and not both for the given metadata

The new XML output has *only* the structured (attribute) method. The plaintext strings are no longer in the descriptions.

The new text output still has plaintext indicators, but they're in separate bullet points instead of part of the description. So they're even less useful to parse than before :)

> finish the pacemaker side sooner than later

I'm hoping we can have it merged next week. The draft was done over a week ago; we've just been reviewing and updating since then. It seems like we're approaching a good place since you and gao were pinged.

> That's easy to fix. Just switch such parameters to `content type="select"` and put the allowed values to option elements, as specified in OCF 1.1 😉

For the most part, they already are and have been for some time now, but `pcs` grabs the "Allowed values: ..." from the `longdesc` instead. For example, in Pacemaker 2.1.7 (latest release):
```
    <parameter name="no-quorum-policy">
      <longdesc lang="en">What to do when the cluster does not have quorum  Allowed values: stop, freeze, ignore, demote, suicide</longdesc>
      <shortdesc lang="en">What to do when the cluster does not have quorum</shortdesc>
      <content type="select" default="">
        <option value="stop" />
        <option value="freeze" />
        <option value="ignore" />
        <option value="demote" />
        <option value="suicide" />
      </content>
    </parameter>
```

I recall there were a couple of options that *should* have been `select`-type with `<content>` and were not until recent commits, but most were.

And yes, it's easy ;) I was just pointing out that I introduced a regression in `pcs` by changing from two spaces to one.

---

> If the old way of running daemon metadata commands is preserved, no matter if it produces metadata in the current format or the one discussed here

In the current proposal, the old way of running daemon metadata commands is preserved, and it produces metadata in the format discussed here.

Reply:
> > Validating against modified OCF schema (`/usr/lib/pcs/data/ocf-1.1.rng`) is hard-coded and it cannot go away.
> 
> All we'd have to do is update that schema to allow two new optional attributes in the `<parameter>` element: `advanced` and `generated`. Since they're optional, it wouldn't break on older Pacemakers.

Agreed, that's the plan. I probably should have stated it clearly: the modified OCF schema isn't going away, but it can be modified as needed.

> > OCF 1.2 standard to differentiate between the two approaches
> 
> I have to defer to Ken on whether it's reasonable to have a new official OCF schema accept this output.

Well, there are more points that would need to be addressed to finally make the OCF schema compatible with agents and pacemaker metadata. Like attributes used by fence agents, custom order of xml elements, percentage and time data types... I don't even remember all the bits which don't comply with the standard and we had to cover them in our custom schema. (It's easy to find them out, though - just take metadata of all agents and try to validate them.) It probably wouldn't pay off to make the new version just for this pacemaker metadata. But having a new version covering _all_ the things actually being used in agents and pacemaker metadata could be worth it.

> > as long as there is either of plaintext or attribute specification and not both for the given metadata
> 
> The new XML output has _only_ the structured (attribute) method. The plaintext strings are no longer in the descriptions.

:+1:

> > finish the pacemaker side sooner than later
> 
> I'm hoping we can have it merged next week. The draft was done over a week ago; we've just been reviewing and updating since then. It seems like we're approaching a good place since you and gao were pinged.

:+1:

> > That's easy to fix. Just switch such parameters to `content type="select"` and put the allowed values to option elements, as specified in OCF 1.1 😉
> 
> They already are. For the most part, they have been for a long time now, but `pcs` still grabs the "Allowed values: ..." from the `longdesc` instead.

Oh, ok. We'll take a look at that. Most probably we want to skip this plaintext parsing in case structured data are provided. We'll do the same for `advanced` and `generated`. That way, pcs would be able to process both the new and the old format.

> > If the old way of running daemon metadata commands is preserved, no matter if it produces metadata in the current format or the one discussed here
> 
> In the current proposal, the old way of running daemon metadata commands is preserved, and it produces metadata in the format discussed here.

:+1:

Reply:
> All `pcs` or `crmsh` would have to do is look for `--list-options` in the output of `crm_attribute --help-all`.

That's what the feature set is for. We don't want external code parsing help

> Well, there are more points that would need to be addressed to finally make the OCF schema compatible with agents and pacemaker metadata. Like attributes used by fence agents, custom order of xml elements, percentage and time data types... I don't even remember all the bits which don't comply with the standard and we had to cover them in our custom schema. (It's easy to find them out, though - just take metadata of all agents and try to validate them.) It probably wouldn't pay off to make the new version just for this pacemaker metadata. But having a new version covering _all_ the things actually being used in agents and pacemaker metadata could be worth it.

Yeah, the "Resource Agent API" column in https://projects.clusterlabs.org/project/view/37/ is collecting possible changes. However resource-agents hasn't even been updated for 1.1 yet, so it would be better to push that forward first.

Reply:
Actually this went into main for 2.1.8. The crm_attribute option is new, so it doesn't matter. The daemons will keep their metadata options for the 2.1 series (maybe even 3.0), and their output will remain the same except it will have the new advanced/generated/deprecated syntax.

Reply:
We could hack it so it keeps the output identical when called via the daemons, but it would be ugly

Reply:
@nrwahl2 , it sounds like it would be better to retrofit it to keep the daemon output identical. We could potentially overload legacy_xml for that purpose.

Reply:
> @nrwahl2 , it sounds like it would be better to retrofit it to keep the daemon output identical.

My thought was that if `advanced="1"`, the high-level tools could add "*** Advanced use only ***" or whatever marker they prefer...

> We could potentially overload legacy_xml for that purpose.

I don't think we should use that any more than we have to. We could add a `bool legacy` option to the new `option-list` message functions, setting it to `false` everywhere except `pcmk__daemon_metadata()`. Keep the legacy stuff clearly separated from the usual format so that it's easy to drop at 3.0.0 or whenever it's allowed.

Reply:
> > @nrwahl2 , it sounds like it would be better to retrofit it to keep the daemon output identical.
> 
> My thought was that if `advanced="1"`, the high-level tools could add "*** Advanced use only ***" or whatever marker they prefer...

Right, but we don't want to require higher-level tools (or user scripts etc.) to change for a minor-minor release. I was hoping they would ignore the new attributes and we'd just lose the markers, but they're actually looking for the markers and/or validating the schema, which means anyone building the newest Pacemaker and running with an existing pcs/crm would have problems.

> > We could potentially overload legacy_xml for that purpose.
> 
> I don't think we should use that any more than we have to. We could add a `bool legacy` option to the new `option-list` message functions, setting it to `false` everywhere except `pcmk__daemon_metadata()` and the fencer metadata. Keep the legacy stuff clearly separated from the usual format so that it's easy to drop at 3.0.0 or whenever it's allowed.

Actually that's why I was thinking of using legacy_xml -- we're already planning to drop that at 3.0, so this would require no extra work then. With legacy_xml we just drop the code that uses it; with a new argument, we'd ideally drop the argument meaning every caller has to be changed.

Reply:
> anyone building the newest Pacemaker and running with an existing pcs/crm would have problems.

Fair enough. I also discovered this morning that `pcs property set` fails now, even with `--force`.

> With legacy_xml we just drop the code that uses it; with a new argument, we'd ideally drop the argument meaning every caller has to be changed.

Makes sense. Honestly the amount of difference in work is negligible. Most of the work will probably be re-implementing the old format alongside the new one, which has to be done for either approach.
* To use `legacy_xml()`, we have to add a getter function so that the message functions can know whether `legacy_xml` is set. The message functions are the ones that would have to actually *use* it. The functions in `output_xml.c` themselves wouldn't do anything special for `option-list` specifically. And the daemon callers would have to set `legacy_xml`. Not a big deal but it's not nothing.
* Adding a new argument involves trivial changes to two or three message functions and only a small handful of callers.

Both of these are pretty easy to add or to drop.

Reply:
Either approach would be fine

Reply:
You win lol. The existence of a nested call chain (`pcmk__list_cluster_options()` and `pcmk__output_cluster_options()`), and the fact that we're going to be adding several more message function callers, makes me think overloading `legacy_xml` is the cleaner approach. It's hackier but we know that it's temporary. I don't foresee there being any weird interference between the two uses, since option listing and `crm_mon` should never overlap.